File: 238-product-of-array-except-self/product-of-array-except-self.cpp
class Solution {
public:
    vector<int> productExceptSelf(vector<int>& nums) {
        vector<int> pro;
        long long int product=1;
        int temp=0;
        for(int i=0; i<nums.size(); i++)
        {
            product*=nums[i];
        }
        for(int i=0; i<nums.size(); i++)
        {
            if(nums[i]!=0)
            {
                temp=product/nums[i];
            }else{
                temp=1;
                for(int j=0; j<nums.size(); j++)
                {
                    if(i!=j)
                    {
                        temp*=nums[j];
                    }
                }
            }
            
            pro.push_back(temp);
        }
        return pro;
    }
};

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        prefix, sufix = [1]*n, [1]*n
        ans = [0]*n
        for i in range(1, n):
            prefix[i] = prefix[i-1] * nums[i-1]
        for i in range(n-2, -1, -1):
            sufix[i] = sufix[i+1]*nums[i+1]
        #print(prefix)
        #print(sufix)
        # [1, 1, 2, 6]
        # [24, 12, 4, 1]
        for i in range(n):
            ans[i] = prefix[i]*sufix[i]
        return ans